### PR TITLE
Restore build_ast in BigQuery backend

### LINKED ISSUE #2717: `build_ast` in BigQuery backend missing
It looks like the `build_ast` function in the BigQuery backend is missing. It causes an error when `.compile()` is called. Here is a repro:

```
import ibis
from ibis.expr.types import TableExpr
import ibis.expr.operations as ops

def _get_ibis_client():
    ibis_client = ibis.backends.bigquery.BigQueryClient
    ibis_client = ibis_client()
    return ibis_client

def _get_test_table(ibis_client):
    names = ['Class',                  
  'end_of_period',          
  'M_Total_Dollars_Total',
            ]
    types = ['string', "date", 'float64']
    schema = ibis.Schema(names, types)
    t = TableExpr(ops.SQLQueryResult("select * from t", schema, ibis_client))
    return t

def test_window_compile():
    ibis_client = _get_ibis_client()
    t = _get_test_table(ibis_client)
    key_cols = ['Class', 'end_of_period']
    order_col = 'end_of_period'
    window = ibis.range_window(
        preceding=ibis.interval(days=62),
        order_by=t[order_col],
        group_by=[t['Class']],
    )
    t = t.mutate(window=t["M_Total_Dollars_Total"].sum().over(window))
    t.compile()
    return t

test_window_compile()
```
```
AttributeError: module 'ibis.backends.bigquery.compiler' has no attribute 'build_ast'
```

Here is a trace of how `.compile()` depends on `build_ast`:
* `SQLClient.compile` calls `SQLClient._build_ast_ensure_limit` ([ibis/client.py#L255](https://github.com/ibis-project/ibis/blob/07f0016abed82be1df5278c29833fd7ab41de0e4/ibis/client.py#L255))
* `SQLClient._build_ast_ensure_limit` calls `SQLClient._build_ast` ([ibis/client.py#L261](https://github.com/ibis-project/ibis/blob/07f0016abed82be1df5278c29833fd7ab41de0e4/ibis/client.py#L261))
* `BigQueryClient._build_ast` tries to call `build_ast` from `ibis/backends/bigquery/compiler.py` ([bigquery/client.py#L439](https://github.com/ibis-project/ibis/blob/165f78de8f4f0121ba2c601b5c9f89bc0f65a593/ibis/backends/bigquery/client.py#L439)) but it doesn't exist

I believe it may have been inadvertently removed in #2661 ([diff](https://github.com/ibis-project/ibis/pull/2661/files#diff-d7c2d45959c2baa00df6631a8818c081adecd87626c52ab82b7560933609e91cL79-L81))

### PR DESCRIPTION
**Pull request by timothydijamco**
This PR fixes #2717 

The code change is just reversing the diff in the lines that are selected here: [Diff of ibis/backends/bigquery/compiler.py from #2661](https://github.com/ibis-project/ibis/pull/2661/files#diff-d7c2d45959c2baa00df6631a8818c081adecd87626c52ab82b7560933609e91cL79-L81)

I was able to confirm locally that `.compile()` using the BigQuery backend works with this PR. In the BigQuery CI, this should make `test_large_compile` in `ibis/backends/bigquery/tests/test_compiler.py` start succeeding again (note: BigQuery CI runs after PRs are merged).

### CODE REVIEW

**Comment by timothydijamco** [comment]
@datapythonista I noticed in some of your backend refactoring changes that you recently added some logic (to the `build_ast` functions in other backends) to create `context` if it doesn't exist:
* [Example diff in the Spark backend](https://github.com/ibis-project/ibis/commit/9fc9ddd5fc8a52abe8fc66f3a2557d7f8070c7d6#diff-604d47455c275f96beef501386f88e6423e326fa94be1f3e85419065509d80b2L49-L51)
* [Example diff in the Impala backend](https://github.com/ibis-project/ibis/commit/9fc9ddd5fc8a52abe8fc66f3a2557d7f8070c7d6#diff-7c0c87e9811a95d5f0f6f2917df3c5de3f7dd78864ef8641a6237f30dfbfba13L19-L22)

Should this be added to BigQuery's `build_ast` as well?

The BigQuery backend seems slightly different because the BigQuery backend's `build_ast` is _only_ called when either of these two functions are called:

1. `SQLClient._build_ast_ensure_limit` ([code](https://github.com/ibis-project/ibis/blob/07f0016abed82be1df5278c29833fd7ab41de0e4/ibis/client.py#L258-L261))
2. `SQLClient.explain` ([code](https://github.com/ibis-project/ibis/blob/07f0016abed82be1df5278c29833fd7ab41de0e4/ibis/client.py#L292-L293))

and both of these functions already ensure `context` exists and is passed to `build_ast`. So right now it seems like `context` will never be `None` when the BigQuery backend's `build_ast` function is called. Because of this, at the least, I don't think restoring `build_ast` like this will have any issues. However I'm open to adding the extra `context`-related logic to `build_ast` if it's something we do want

**Comment by jreback** [comment]
@timothydijamco are we running the BQ tests in the CI?

**Comment by jreback** [comment]
actually @timothydijamco i think you mentioned we run the BQ tests only on master (e.g. after merge).

however is it possible to test some translations directly to avoid changes like this

maybe a more complicated question, cc @tswast

**Comment by timothydijamco** [comment]
@jreback Yes the BigQuery tests run on master only. Do you know why? (Is it related to needing the Ibis BigQuery credentials?)

I believe `.compile()` by itself doesn't access/require any BigQuery credentials, project ID, etc and so theoretically BigQuery backend compilations _could_ be tested without needing any of those things. However, I think with way the BigQuery backend tests are currently configured, the test data for all the tests comes from (i.e. requires) Ibis's BigQuery project. (@tswast let me know if any of this doesn't sound right)

**Comment by jreback** [comment]
> @jreback Yes the BigQuery tests run on master only. Do you know why? (Is it related to needing the Ibis BigQuery credentials?)
> 
> I believe `.compile()` by itself doesn't access/require any BigQuery credentials, project ID, etc and so theoretically BigQuery backend compilations _could_ be tested without needing any of those things. However, I think with way the BigQuery backend tests are currently configured, the test data for all the tests comes from (i.e. requires) Ibis's BigQuery project. (@tswast let me know if any of this doesn't sound right)

this is correct, we could compare .compile() outputs (prob not for everything), but yeah its a credential issue.

can you add a compile test for this? (that fails on current but passes with this)?

**Comment by datapythonista** [comment]
Thanks for this @timothydijamco 

Yes, as you guessed, BigQuery tests connect to an actual BigQuery instance, and the credentials (GitHub secrets) are only available after a PR is merged (so, commits to `master`), otherwise a malicious PR could access them.

Ideally I think BigQuery tests should use a mock BigQuery instance, and then have a specific set of tests that run with an actual BigQuery instance. Something easier to do in the short term is to download the credentials and dependencies of bigquery locally, and test changes locally.

I personally think that the way to move forward is moving the backend to a separate repo (see #2665). We have currently 14 backends in this repo, and I don't think it's possible that for each PR here, we run the tests locally for BigQuery. Also, BigQuery tests have always been broken since I joined the project more than a year ago. So, after any commit in `master` we get a failure from the BigQuery build (see #2353).

I was thinking on moving the backend to a separate repo after the backend refactoring I'm doing. But since it's difficult to make sure we don't break things on it, I think it's probably better to wait until the refactoring is finished, and we release Ibis 2.0, and then change the BigQuery backend API to adapt to the new changes. We'll do this with the Omnisci backend (and mssql and any other too).

@tswast @jreback what do you think about moving the backend to a repo in a Google organization now?

**Comment by jreback** [comment]
@datapythonista that sounds fine. for this PR @timothydijamco can you add a test in here: https://github.com/ibis-project/ibis/blob/master/ibis/backends/bigquery/tests/test_compiler.py

these should be able to run w/o the actual backend I think

**Comment by tswast** [comment]
Unfortunately, even a bunch of the tests in `test_compiler` require a connection to BigQuery, as they depend on the existence of particular tables. Also, we don't have an emulator for BigQuery, so we're kind of stuck without additional CI work to safely allow use of secrets only after review (which is probably best done in the final repo location)

**Comment by datapythonista** [comment]
Thanks for the info @tswast. Do you think we can move the backend to the new repo already?

**Comment by tswast** [comment]
> Do you think we can move the backend to the new repo already?

I'm open to that. I can dedicate some time to getting the credentials setup after the move.

**Comment by datapythonista** [comment]
Amazing. In case it's useful, this is the PR I did to move the Omnisci backend: https://github.com/omnisci/ibis-omniscidb/pull/1/files

Let me know if you need any help.

**Review by jreback — APPROVED** [review]
ok merging, hopefully will pass tests

**Comment by jreback** [comment]
thanks @timothydijamco

**Comment by timothydijamco** [comment]
> @datapythonista that sounds fine. for this PR @timothydijamco can you add a test in here: https://github.com/ibis-project/ibis/blob/master/ibis/backends/bigquery/tests/test_compiler.py
> 
> these should be able to run w/o the actual backend I think

> Unfortunately, even a bunch of the tests in `test_compiler` require a connection to BigQuery, as they depend on the existence of particular tables. Also, we don't have an emulator for BigQuery, so we're kind of stuck without additional CI work to safely allow use of secrets only after review (which is probably best done in the final repo location)

Sorry for the late response here—related to what @tswast was saying I think adding a new BigQuery compile test that is runnable before merging would have probably also required some changes in how the BigQuery backend test suite is set up

However now that this PR is merged and the BigQuery CI ran, I can confirm `test_large_compile` is no longer failing:
- [BigQuery backend CI run for preceding PR](https://github.com/ibis-project/ibis/runs/2254534009?check_suite_focus=true): `FAILED ibis/backends/bigquery/tests/test_compiler.py::test_large_compile`
- [BigQuery backend CI run for this PR](https://github.com/ibis-project/ibis/runs/2296596811?check_suite_focus=true): No failure for `test_large_compile`

**Pull request merged**
